src/IO/H5/VolumeData.cpp

```diff
@@ -154,6 +162,308 @@ void append_element_extents_and_connectivity(
+    h_ref_array[i] = h_ref;
+    additional_connectivity_length *= pow(2, h_ref + 1) - 1;
+    h_ref_previous_start_position = h_ref_start_position;
+  }
+
+  expected_connectivity_length +=
+      (additional_connectivity_length - block_extents.size()) * 8;
+
+  return std::tuple{expected_connectivity_length,
+                    expected_number_of_grid_points, h_ref_array};
+}
+
+// Generates the mesh
+template <size_t SpatialDim>
+Mesh<SpatialDim> generate_element_mesh(
+    const std::vector<std::string>& element_bases,
+    const std::vector<std::string>& element_quadratures,
+    const std::vector<size_t>& element_extents) {
+  std::array<Spectral::Basis, SpatialDim> basis_array = {};
+
+  std::array<Spectral::Quadrature, SpatialDim> quadrature_array = {};
+
+  std::array<size_t, SpatialDim> extents_array = {};
+
+  for (size_t i = 0; i < SpatialDim; i++) {
+    basis_array[i] = Spectral::to_basis(element_bases[i]);
+    quadrature_array[i] = Spectral::to_quadrature(element_quadratures[i]);
+    extents_array[i] = element_extents[i];
+  }
+
+  return Mesh<SpatialDim>{extents_array, basis_array, quadrature_array};
+}
+
+template <size_t SpatialDim>
+std::vector<std::array<double, SpatialDim>> generate_block_logical_coordinates(
+    const std::vector<std::array<double, SpatialDim>>&
+        element_logical_coordinates,
+    const std::string& grid_name,
+    const std::array<int, SpatialDim>& h_refinement_array) {
+  size_t grid_points_x_start_position = 0;
+  std::vector<std::array<double, SpatialDim>> block_logical_coordinates;
+  block_logical_coordinates.reserve(element_logical_coordinates.size());
+  std::vector<double> number_of_elements_each_direction;
+  number_of_elements_each_direction.reserve(SpatialDim);
+  std::vector<double> shift_each_direction;
+  shift_each_direction.reserve(SpatialDim);
+
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    double number_of_elements = pow(2, h_refinement_array[i]);
+    number_of_elements_each_direction.push_back(number_of_elements);
+    size_t grid_points_start_position =
+        grid_name.find('I', grid_points_x_start_position + 1);
+    size_t grid_points_end_position =
+        grid_name.find(',', grid_points_start_position);
+    if (i == SpatialDim) {
+      grid_points_end_position =
+          grid_name.find(')', grid_points_start_position);
+    }
+    int element_index = std::stoi(grid_name.substr(
+        grid_points_start_position + 1,
+        grid_points_end_position - grid_points_start_position - 1));
+    double shift = (-1 + (2 * element_index + 1) / number_of_elements);
+    shift_each_direction.push_back(shift);
+    grid_points_x_start_position = grid_points_start_position;
+  }
+
+  for (size_t i = 0; i < element_logical_coordinates.size(); ++i) {
+    std::array<double, SpatialDim> grid_point_coordinate = {};
+    for (size_t j = 0; j < grid_point_coordinate.size(); ++j) {
+      grid_point_coordinate[j] = 1 / number_of_elements_each_direction[j] *
```

Comment:
do you want floating point or integer division here? If integer, it's important to add that as a comment!

Reply:
After looking at it closer, I think it will always be floating point division. Is the expected change then 

`1/blah` --> `1./blah`

Reply:
Yep!

Reply:
ping?